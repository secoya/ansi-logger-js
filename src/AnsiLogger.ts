import * as clc from 'cli-color';
import * as moment from 'moment';
import { IdentityTransformer } from './IdentityTransformer';
import { JSONTransformer } from './JSONTransformer';
import { TextTransformer } from './TextTransformer';

export interface LogEntry {
	group?: string;
	levelNumeric: number;
	levelText: string | null;
	message: string | null;
	timestamp: string;
}

export type Transformer = (this: AnsiLogger, entry: LogEntry) => any;

function assertNever(__: never, msg?: string) {
	throw new Error(msg || 'Unsupported option ' + __);
}

// tslint:disable:no-bitwise
export enum Mask {
	ERROR = 0b0000001,
	WARN = ERROR << 1,
	SUCCESS = WARN << 1,
	LOG = SUCCESS << 1,
	INFO = LOG << 1,
	DEBUG = INFO << 1,
	VERBOSE = DEBUG << 1,
}

export enum Level {
	SILENT = 0b0000000,
	ERROR = SILENT | Mask.ERROR,
	WARN = ERROR | Mask.WARN,
	SUCCESS = WARN | Mask.SUCCESS,
	LOG = SUCCESS | Mask.LOG,
	INFO = LOG | Mask.INFO,
	DEBUG = INFO | Mask.DEBUG,
	VERBOSE = DEBUG | Mask.VERBOSE,
}
// tslint:enable:no-bitwise

export interface LoggerOptionsInput {
	colors: { [color: string]: clc.Format };
	group: string;
	'group-color': clc.Format;
	'log-level': number;
	'no-colors': boolean;
	outputters: {
		err: (msg: any) => void;
		out: (msg: any) => void;
	};
	'startup-info': boolean;
	timeformat: string;
	transformer: (this: AnsiLogger, entry: LogEntry) => any;
}

export interface LoggerOptions {
	group?: LoggerOptionsInput['group'];
	'group-color'?: LoggerOptionsInput['group-color'];

	'log-level': LoggerOptionsInput['log-level'];
	'no-colors': LoggerOptionsInput['no-colors'];
	outputters: LoggerOptionsInput['outputters'];
	'startup-info': LoggerOptionsInput['startup-info'];
	timeformat: LoggerOptionsInput['timeformat'];
	transformer: LoggerOptionsInput['transformer'];
}

export type LogMask = keyof typeof Mask;
export type LogLevel = keyof typeof Level;

export type ColorMap = { [mask: number]: clc.Format } & { TIME: clc.Format } & { TITLE: clc.Format };

export function matchMask(level: number, mask: number): boolean {
	// tslint:disable-next-line:no-bitwise
	return (level & mask) === mask;
}

const knownColors = Object.keys(Level).concat(['TIME', 'TITLE']);
function isKnownColorType(color: string): color is LogLevel | 'TIME' | 'TITLE' {
	return knownColors.includes(color);
}

/**
 * Ansi output logger.
 * This controls what should be ouputted to the console,
 * everything is categorized into log levels, so when you set a log level
 * you output from all the selected levels.
 * It is possible to disables colors (some teminals don't support colors).
 * you can also specify that you are only interested in output for a specific
 * log-level, then everything else is not outputted.
 * It is also possible to make the logger silent.
 */
export class AnsiLogger {
	// tslint:enable:react-aware-member-ordering

	public get options(): Partial<LoggerOptions> {
		return this._options;
	}

	/**
	 * The options object holder.
	 * This is filled with the default values when the Logger is constructed,
	 * it can be changed by using the setOptions method.
	 */
	// tslint:disable-next-line:variable-name react-aware-member-ordering
	private _options: LoggerOptions;
	/**
	 * Default color scheme
	 * can be overriden by providing a
	 * new full or partial ColorMap.
	 */
	public readonly colors: ColorMap = {
		[Mask.ERROR]: clc.bgRed.white,
		[Mask.WARN]: clc.red.bold,
		[Mask.SUCCESS]: clc.green,
		[Mask.LOG]: clc,
		[Mask.INFO]: clc.blue,
		[Mask.DEBUG]: clc.yellow,
		[Mask.VERBOSE]: clc.magenta,
		TIME: clc.cyan,
		TITLE: clc.cyan,
	};

	/**
	 * Constructs a Logger, and sets default option values.
	 */
	public constructor(options?: Partial<LoggerOptionsInput>) {
		// tslint:disable:object-literal-key-quotes
		this._options = {
			group: undefined,
			'group-color': undefined,
			'log-level': Level.INFO,
			'no-colors': false,
			outputters: {
				out(msg: any): void {
					// Setting up default group // the log level // disbles colors if true
					process.stdout.write(msg + '\n');
				},
				err(msg: any): void {
					process.stderr.write(msg + '\n');
				},
			},
			'startup-info': true,
			/**
			 * Moment.js formats.
			 * @link http://momentjs.com
			 */
			timeformat: 'YYYY-MM-DD\\THH:mm:ss.SSSZZ',
			transformer: TextTransformer,
		};
		// tslint:enable:object-literal-key-quotes

		if (options != null) {
			this.setOptions(options);
		}
	}

	/**
	 * Print to console.
	 * NB! This will print to the console based on the log level you have enabled.
	 */
	private print(msg: string, logMask?: number | null): void {
		// return if the log-level is higher than the selected the log-level.
		if (logMask == null) {
			logMask = Mask.LOG;
		}

		if (!matchMask(this._options['log-level'], logMask)) {
			return;
		}

		const entry: LogEntry = {
			group: this.options.group,
			levelNumeric: logMask,
			levelText: this.resolveLogLevel(logMask),
			message: msg,
			timestamp: moment().format(this.options.timeformat),
		};

		// finally printing the output!.
		if (matchMask(logMask, Mask.ERROR)) {
			this._options.outputters.err.call(this, this._options.transformer.call(this, entry));
		} else {
			this._options.outputters.out.call(this, this._options.transformer.call(this, entry));
		}
	}

	/**
	 * Setting a single color in the `ColorMap`
	 */
	private setColor(level: LogLevel | 'TITLE' | 'TIME', color: clc.Format) {
		switch (level) {
			case 'ERROR':
				this.colors[Mask.ERROR] = color;
				break;
			case 'WARN':
				this.colors[Mask.WARN] = color;
				break;
			case 'SUCCESS':
				this.colors[Mask.SUCCESS] = color;
				break;
			case 'LOG':
				this.colors[Mask.LOG] = color;
				break;
			case 'INFO':
				this.colors[Mask.INFO] = color;
				break;
			case 'DEBUG':
				this.colors[Mask.DEBUG] = color;
				break;
			case 'VERBOSE':
				this.colors[Mask.VERBOSE] = color;
				break;
			case 'TITLE':
				this.colors.TITLE = color;
				break;
			case 'TIME':
				this.colors.TIME = color;
				break;
			case 'SILENT':
				break;
			default:
				assertNever(level);
				break;
		}
	}

	/**
	 * Print a debug formatted message.
	 */
	public debug<T>(firstArg: T, ...__: any[]): T {
		for (const msg of Array.from(arguments)) {
			if (typeof msg === 'string') {
				this.print(msg, Mask.DEBUG);
			} else {
				this.print(this.formatTypes(msg), Mask.DEBUG);
			}
		}
		return firstArg;
	}

	/**
	 * Print an error.
	 */
	public error<T>(firstArg: T, ...__: any[]): T {
		for (const msg of Array.from(arguments)) {
			if (msg instanceof Error || (typeof msg === 'object' && 'stack' in msg)) {
				this.formatError(msg);
			} else {
				this.print(msg, Mask.ERROR);
			}
		}
		return firstArg;
	}

	/**
	 * Format an error, this is typically used, for string formatting an Exception/Error.
	 */
	public formatError(err: any): void {
		this.print(this.formatTypes(err), Mask.ERROR);
	}

	/**
	 * Format a function call, for the debug level
	 * NB! if passing arguments the function
	 *     every argument, gonna be formatted with the formatTypes() function
	 */
	public formatFunctionCall(functionName: string, args?: any[]) {
		if (args == null) {
			args = [];
		}
		const formattedArgs = [];
		for (const a of Array.from(args)) {
			formattedArgs.push(this.formatTypes(a));
		}
		return this.debug(functionName + `(${formattedArgs.join(', ')})`);
	}

	/**
	 * Format types to string, some types make resively calls.
	 */
	// tslint:disable-next-line:no-reserved-keywords
	public formatTypes(type: any, depth?: number, indent?: number) {
		if (this.options.transformer === JSONTransformer || this.options.transformer === IdentityTransformer) {
			return type;
		}

		// making the proper indentation
		let val;
		if (indent == null) {
			indent = 0;
		}
		if (depth == null) {
			depth = 3;
		}

		const pad = ' '.repeat(indent > 0 ? indent - 1 : 0);

		// primitive types
		if (typeof type === 'number' || typeof type === 'boolean' || type == null) {
			return `${pad}${type}`;
		}

		if (typeof type === 'string') {
			return `${pad}'${type}'`;
		}

		// array is formatted as one-liners
		if (Array.isArray(type) && indent < depth) {
			let str = `${pad}[`;
			// tslint:disable-next-line:forin
			for (const key of type.keys()) {
				val = type[key];
				str += ` ${this.formatTypes(val, indent + 1, depth).trim()}`;
				str += ((key as any) as number) < type.length - 1 ? ',' : ' ';
			}
			return `${str}]`;
		}

		// hashes is formatted with indentation for every level
		// of the object, the values of the properties are also resolved.
		if (typeof type === 'object') {
			const cname = type.constructor.name;
			if (cname === 'Error' || type instanceof Error) {
				let str = `${pad}${type.message}`;
				if (type.stack != null) {
					str += `\n${pad}${type.stack}`;
				}
				return str;
			} else if (cname === 'Object' && indent < depth) {
				let str = `${pad}{`;
				for (const key of Object.keys(type)) {
					val = type[key];
					str += `\n${pad}  ${key}: ${this.formatTypes(val, indent + 1, depth).trim()}`;
				}
				if (str.length > pad.length + 1) {
					str += `\n${pad}`;
				}
				return `${str}}`;
			}
		}

		// print the name of a complex type.
		const isFunc = typeof type === 'function';
		if (isFunc) {
			const cname = type.constructor.name;
			if (isFunc && cname !== 'Function') {
				return `${pad}[Function: ${cname}]`;
			}
			return `${pad}${cname}`;
		}

		// print the name of an unhandled type.
		// typically this will return 'object'
		return `${pad}${typeof type}`;
	}

	/**
	 * Print an info formatted message.
	 */
	public info<T>(firstArg: T, ...__: any[]): T {
		for (const msg of Array.from(arguments)) {
			if (typeof msg === 'string') {
				this.print(msg, Mask.INFO);
			} else {
				this.print(this.formatTypes(msg), Mask.INFO);
			}
		}
		return firstArg;
	}

	/**
	 * Print an default log formatted message.
	 */
	public log<T>(firstArg: T, ...__: any[]): T {
		for (const msg of Array.from(arguments)) {
			if (typeof msg === 'string') {
				this.print(msg, Mask.LOG);
			} else {
				this.print(this.formatTypes(msg), Mask.LOG);
			}
		}
		return firstArg;
	}

	/**
	 * Resolve a string representation of the log-level.
	 */
	public resolveLogLevel(mask: number) {
		switch (mask) {
			case Mask.ERROR:
				return 'ERROR';
			case Mask.WARN:
				return 'WARN';
			case Mask.SUCCESS:
				return 'SUCCESS';
			case Mask.INFO:
				return 'INFO';
			case Mask.DEBUG:
				return 'DEBUG';
			case Mask.VERBOSE:
				return 'VERBOSE';
			case Mask.LOG:
				return 'LOG';
			default:
				return 'CUSTOM';
		}
	}

	/**
	 * Set new colors.
	 */
	public setColors(colorMap: { [level: string]: clc.Format }) {
		for (const level of Object.keys(colorMap)) {
			const color = colorMap[level];
			const needle = level.toUpperCase();
			if (isKnownColorType(needle)) {
				this.setColor(needle, color);
			}
		}
	}

	/**
	 * Sets the options.
	 */
	public setOptions(options: Partial<LoggerOptionsInput>) {
		if (options.colors != null) {
			this.setColors(options.colors);
		}
		const currentLoglevel = this.options['log-level'];
		const optionKeys = Object.keys(this.options);
		for (const key of optionKeys) {
			const val = (options as any)[key];
			if (val != null) {
				(this.options as any)[key] = val;
			}
		}

		if (!Number.isInteger(this._options['log-level']) || this._options['log-level'] > Level.VERBOSE) {
			this.options['log-level'] = currentLoglevel;
			this.warn(`Invalid log level is trying to be set: ${options['log-level']}, aborting...`);
		}
	}

	/**
	 * Print a success formatted message.
	 */
	public success<T>(firstArg: T, ...__: any[]): T {
		for (const msg of Array.from(arguments)) {
			this.print(msg, Mask.SUCCESS);
		}
		return firstArg;
	}

	/**
	 * Print a title formatted message.
	 */
	public title<T>(msg: T, ...__: any[]): T {
		for (const m of Array.from(arguments)) {
			this.print(m, Mask.INFO);
		}
		return msg;
	}

	/**
	 * Print a verbose formatted message.
	 */
	public verbose<T>(firstArg: T, ...__: any[]): T {
		for (const msg of Array.from(arguments)) {
			this.print(msg, Mask.VERBOSE);
		}
		return firstArg;
	}

	/**
	 * Print a warning.
	 */
	public warn<T>(firstArg: T, ...__: any[]): T {
		for (const msg of Array.from(arguments)) {
			this.print(msg, Mask.WARN);
		}
		return firstArg;
	}
}

export default AnsiLogger;
