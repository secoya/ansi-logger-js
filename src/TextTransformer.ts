import * as clc from 'cli-color';
import { matchMask, resolveLogLevel, LogEntry, LogLevel, Mask, Transformer } from './AnsiLogger';

function assertNever(__: never, msg?: string) {
	throw new Error(msg || 'Unsupported option ' + __);
}

const knownColors = ['SILENT', 'ERROR', 'WARN', 'SUCCESS', 'LOG', 'INFO', 'DEBUG', 'VERBOSE'].concat(['TIME', 'TITLE']);
function isKnownColorType(color: string): color is LogLevel | 'TIME' | 'TITLE' {
	return knownColors.includes(color);
}

export type ColorMap = { [mask: number]: clc.Format } & { TIME: clc.Format } & { TITLE: clc.Format };

export interface TextTransformerOptionsInput {
	colorMap: { [color: string]: clc.Format };
	colors: boolean;
	groupColor: clc.Format;
}

export interface TextTransformerOptions {
	colors: TextTransformerOptionsInput['colors'];
	groupColor?: clc.Format;
}

export class TextTransformer implements Transformer {
	public get options(): Partial<TextTransformerOptions> {
		return this._options;
	}

	// tslint:disable-next-line:variable-name
	private _options: TextTransformerOptions;

	/**
	 * Default color scheme
	 * can be overriden by providing a
	 * new full or partial ColorMap.
	 */
	public readonly colors: ColorMap = {
		[Mask.ERROR]: clc.bgRed.white,
		[Mask.WARN]: clc.red.bold,
		[Mask.SUCCESS]: clc.green,
		[Mask.LOG]: clc,
		[Mask.INFO]: clc.blue,
		[Mask.DEBUG]: clc.yellow,
		[Mask.VERBOSE]: clc.magenta,
		TIME: clc.cyan,
		TITLE: clc.cyan,
	};

	public constructor(options?: Partial<TextTransformerOptionsInput>) {
		this._options = {
			colors: true,
		};

		if (options != null) {
			this.setOptions(options);
		}
	}

	/**
	 * Colorize the message string.
	 * NB! If no-colors mode is on or no color is given.
	 * then this method just return the message as it is.
	 */
	private colorize(msg: string, color?: clc.Format | null) {
		if (color == null || !process.stdout.isTTY || !this._options.colors) {
			return msg;
		}
		return color(msg);
	}

	/**
	 * Format group if any.
	 */
	private formatGroup(group: string) {
		const groupTrimmed = group.trim();
		const pad = groupTrimmed.length - group.length;
		let padding = '';
		for (const __ of Array(pad)) {
			padding += ' ';
		}
		return `[${this.colorize(groupTrimmed, this.options.groupColor)}]${padding}`;
	}

	/**
	 * Format the loglevel to the console
	 */
	private formatLogLevel(loglevel: number) {
		// no need to ouput the log level, if the default log level is selected.
		// then it's just a waste of space.
		const loglevelStr = resolveLogLevel(loglevel);
		const pad = ' '.repeat(loglevelStr.length < 6 ? 7 - loglevelStr.length : 0);

		// resolving the color for the log level.
		const loglevelColor = this.resolveLevelColor(loglevel);

		// the formatted log-level
		return `[${this.colorize(loglevelStr, loglevelColor)}]${pad}`;
	}

	/**
	 * Format a object to string.
	 */
	private formatTime(time: string) {
		return `[${this.colorize(time, this.colors.TIME)}]`;
	}

	private handleMultiline(msg: string, color?: clc.Format | null): string {
		const res = [];
		// colorize each line, so when the string is splitted later,
		// it will not mess up the colors.
		if ((msg != null ? msg.split : undefined) == null) {
			msg = `${msg}`;
		}
		for (const m of Array.from(msg.split('\n'))) {
			res.push(this.colorize(m, color));
		}
		return res.join('\n');
	}

	/**
	 *
	 * @param level
	 */
	private resolveLevelColor(level: number): clc.Format {
		switch (true) {
			case matchMask(level, Mask.ERROR):
				return this.colors[Mask.ERROR];
			case matchMask(level, Mask.WARN):
				return this.colors[Mask.WARN];
			case matchMask(level, Mask.SUCCESS):
				return this.colors[Mask.SUCCESS];
			case matchMask(level, Mask.INFO):
				return this.colors[Mask.INFO];
			case matchMask(level, Mask.DEBUG):
				return this.colors[Mask.DEBUG];
			case matchMask(level, Mask.VERBOSE):
				return this.colors[Mask.VERBOSE];
			default:
				return this.colors[Mask.LOG];
		}
	}

	/**
	 * Setting a single color in the `ColorMap`
	 */
	private setColor(level: LogLevel | 'TITLE' | 'TIME', color: clc.Format) {
		switch (level) {
			case 'ERROR':
				this.colors[Mask.ERROR] = color;
				break;
			case 'WARN':
				this.colors[Mask.WARN] = color;
				break;
			case 'SUCCESS':
				this.colors[Mask.SUCCESS] = color;
				break;
			case 'LOG':
				this.colors[Mask.LOG] = color;
				break;
			case 'INFO':
				this.colors[Mask.INFO] = color;
				break;
			case 'DEBUG':
				this.colors[Mask.DEBUG] = color;
				break;
			case 'VERBOSE':
				this.colors[Mask.VERBOSE] = color;
				break;
			case 'TITLE':
				this.colors.TITLE = color;
				break;
			case 'TIME':
				this.colors.TIME = color;
				break;
			case 'SILENT':
				break;
			default:
				assertNever(level);
				break;
		}
	}

	private setOptions(options: Partial<TextTransformerOptionsInput>) {
		if (options.colorMap != null) {
			this.setColors(options.colorMap);
		}

		const optionKeys = Object.keys(this.options) as (keyof TextTransformerOptions)[];
		for (const key of optionKeys) {
			const val = options[key];
			if (val != null) {
				this._options[key] = val;
			}
		}
	}

	/**
	 * Transform log entry to text output.
	 */
	public format(entry: LogEntry): string {
		// get the formatted current time.
		let str = this.formatTime(entry.timestamp);

		// get the formatted group
		if (entry.group != null) {
			str += ` ${this.formatGroup(entry.group)}`;
		}

		// get the formatted log-level.
		if (entry.levelText != null) {
			const levelText = this.formatLogLevel(entry.levelNumeric);
			if (levelText != null) {
				str += ` ${levelText}`;
			}
		}

		// now insert the time and log-level on each line.
		if (entry.message != null) {
			str += ` ${entry.message.replace(/\n/g, `\n${str} `)}`;
		}

		return this.handleMultiline(str, this.colors[entry.levelNumeric]);
	}

	/**
	 * Set new colors.
	 */
	public setColors(colorMap: { [level: string]: clc.Format }) {
		for (const level of Object.keys(colorMap)) {
			const color = colorMap[level];
			const needle = level.toUpperCase();
			if (isKnownColorType(needle)) {
				this.setColor(needle, color);
			}
		}
	}
}
