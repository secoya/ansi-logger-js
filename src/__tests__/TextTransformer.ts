import { resolveLogLevel, AnsiLogger, LogEntry, Mask } from '../AnsiLogger';
import { ColorMap, TextTransformer } from '../TextTransformer';

describe('TextTransformer', () => {
	describe('printer default', () => {
		const orgStdoutWrite = process.stdout.write;
		const orgStderrWrite = process.stderr.write;

		afterEach(() => {
			process.stderr.write = orgStderrWrite;
			process.stdout.write = orgStdoutWrite;
		});

		test('if no printer is passed to constructor stdout and stderr is used', () => {
			const transformer = new TextTransformer();
			const logger = new AnsiLogger({
				transformer: transformer,
			});

			const stderrWrite = jest.fn();
			const stdoutWrite = jest.fn();

			process.stderr.write = stderrWrite;
			process.stdout.write = stdoutWrite;

			logger.error('error');
			logger.info('info');

			expect(stderrWrite).toHaveBeenCalledTimes(1);
			expect(stdoutWrite).toHaveBeenCalledTimes(1);
		});
	});

	describe('custom printer', () => {
		test('provided printer is used', () => {
			const err = jest.fn();
			const out = jest.fn();

			const transformer = new TextTransformer({
				printer: {
					err: err,
					out: out,
				},
			});

			const logger = new AnsiLogger({
				transformer: transformer,
			});

			logger.error('error');
			logger.info('info');

			expect(err).toHaveBeenCalledTimes(1);
			expect(out).toHaveBeenCalledTimes(1);
		});
	});

	describe('colors', () => {
		const orgIsTTY = process.stdout.isTTY;
		afterEach(() => {
			process.stdout.isTTY = orgIsTTY;
		});

		test('enable colors option, results in colorizing is used if prints to a TTY', () => {
			const err = jest.fn();
			const out = jest.fn();

			process.stdout.isTTY = undefined;
			const transformer = new TextTransformer({
				colors: true,
				printer: {
					err: err,
					out: out,
				},
			});

			expect(transformer.useColors).toBe(false);
			process.stdout.isTTY = true;
			expect(transformer.useColors).toBe(true);
		});

		test('disable colors option, results in colors are not used, even when prints to a TTY', () => {
			const err = jest.fn();
			const out = jest.fn();

			process.stdout.isTTY = true;
			const transformer = new TextTransformer({
				colors: false,
				printer: {
					err: err,
					out: out,
				},
			});

			expect(transformer.useColors).toBe(false);
		});

		// tslint:disable-next-line:max-line-length
		test('force colors option, results in colors are always used, even when printing to a non tty and colors option is disabled', () => {
			const err = jest.fn();
			const out = jest.fn();

			process.stdout.isTTY = undefined;
			const transformer = new TextTransformer({
				colors: false,
				forceColors: true,
				printer: {
					err: err,
					out: out,
				},
			});

			expect(transformer.useColors).toBe(true);

			expect(transformer.format({
				group: 'text',
				levelNumeric: Mask.INFO,
				levelText: resolveLogLevel(Mask.INFO),
				message: 'info',
				timestamp: 'NOW',
			})).toMatchSnapshot();
		});

		test('setting color via options', () => {
			// tslint:disable:object-literal-sort-keys
			const colorMap: ColorMap = {
				ERROR: jest.fn(),
				WARN: jest.fn(),
				SUCCESS: jest.fn(),
				LOG: jest.fn(),
				INFO: jest.fn(),
				DEBUG: jest.fn(),
				VERBOSE: jest.fn(),
				GROUP: jest.fn(),
				TIME: jest.fn(),
			} as any;

			// tslint:enable:object-literal-sort-keys
			const transformer = new TextTransformer({
				colorMap: colorMap,
			});

			expect(transformer.colors.ERROR).toBe(colorMap.ERROR);
			expect(transformer.colors.WARN).toBe(colorMap.WARN);
			expect(transformer.colors.SUCCESS).toBe(colorMap.SUCCESS);
			expect(transformer.colors.LOG).toBe(colorMap.LOG);
			expect(transformer.colors.INFO).toBe(colorMap.INFO);
			expect(transformer.colors.DEBUG).toBe(colorMap.DEBUG);
			expect(transformer.colors.VERBOSE).toBe(colorMap.VERBOSE);
			expect(transformer.colors.GROUP).toBe(colorMap.GROUP);
			expect(transformer.colors.TIME).toBe(colorMap.TIME);
		});

		test('setting invalid coloring function throws', () => {
			const colorMap = {
				INVALID: () => { return; },
			} as any;
			expect(() => {
				return new TextTransformer({
					colorMap: colorMap,
				});
			}).toThrowErrorMatchingSnapshot();
		});
	});

	describe('format', () => {
		test('returns string', () => {
			const transformer = new TextTransformer();
			const message = null;
			const logEntry: LogEntry = {
				group: 'text',
				levelNumeric: Mask.INFO,
				levelText: resolveLogLevel(Mask.INFO),
				message: message,
				timestamp: 'NOW',
			};
			expect(transformer.format(logEntry)).toEqual(expect.any(String));
		});

		test('formatting errors without stack does not fail', () => {
			const transformer = new TextTransformer();
			const message = new Error('Error: no stack');
			(message as any).stack = null;
			const logEntry: LogEntry = {
				group: 'text',
				levelNumeric: Mask.ERROR,
				levelText: resolveLogLevel(Mask.ERROR),
				message: message,
				timestamp: 'NOW',
			};

			expect(transformer.format(logEntry)).toMatchSnapshot();
		});

		test('complex types', () => {
			const err = jest.fn();
			const out = jest.fn();

			const logger = new AnsiLogger({
				timeformat: '0000-00-00 00:00:00.000+0000',
				transformer: new TextTransformer({
					printer: {
						err: err,
						out: out,
					},
				}),
			});

			const complexMessage = {
				featureEnabled: false,
				fn: () => { return; },
				list: ['1', '2', '3'],
				nested: {
					level: {
						must: {
							be: {
								exceeded: true,
							},
						},
					},
				},
				noValue: {
					provided: null,
					unprovided: undefined,
				},
				numeric: Math.PI,
				string: 'complex',
				transformer: new TextTransformer(),
			};

			logger.info(complexMessage);

			expect(out.mock.calls[0][0]).toMatchSnapshot();
		});

		test('custom level text mask, will simply output get outputtet as is', () => {
			const transformer = new TextTransformer();
			const message = 'message';
			const logEntry: LogEntry = {
				group: 'text',
				levelNumeric: Mask.INFO,
				levelText: 'CUSTOM',
				message: message,
				timestamp: 'NOW',
			} as any;

			expect(transformer.format(logEntry)).toMatchSnapshot();
		});

		test('unknown numeric level mask, without custom level text, will output UNKNOWN', () => {
			const transformer = new TextTransformer();
			const message = 'message';
			const logEntry: LogEntry = {
				group: 'text',
				levelNumeric: Mask.INFO,
				message: message,
				timestamp: 'NOW',
			} as any;

			expect(transformer.format(logEntry)).toMatchSnapshot();
		});

	});

	describe('formatTypes', () => {
		test('returnes a string', () => {
			const complexMessage = {
				some: 'complex',
				structure: [1, true],
			};
			const transformer = new TextTransformer();
			expect(transformer.formatTypes(complexMessage)).toEqual(expect.any(String));
		});

		test('named functions get outputted with name suffix', () => {
			const transformer = new TextTransformer();
			function withName() {
				return;
			}

			const message = {
				fns: [() => { return; }],
				withName: withName,
			};

			expect(transformer.formatTypes(message)).toMatchSnapshot();
		});

		test('formatting errors without stack without breaking', () => {
			const transformer = new TextTransformer();
			const errorWithoutStack = new Error('Without stack');
			(errorWithoutStack as any).stack = null;
			const errorWithStack = new Error('With stack');
			(errorWithStack as any).stack = `Error: With stack
at ... Mock stack trace`;
			const message = {
				errors: [
					errorWithoutStack,
					errorWithStack,
				],
			};

			expect(transformer.formatTypes(message)).toMatchSnapshot();
		});
	});
});
