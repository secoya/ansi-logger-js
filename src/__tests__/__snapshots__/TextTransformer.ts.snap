// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextTransformer colors force colors option, results in colors are always used, even when printing to a non tty and colors option is disabled 1`] = `
"[[36mNOW[39m] [[33mtext[39m] [[34mINFO[39m]    [34minfo[39m
"
`;

exports[`TextTransformer colors setting invalid coloring function throws 1`] = `"Unsupported option INVALID"`;

exports[`TextTransformer format complex types 1`] = `
"[0000-00-00 00:00:00.000+0000] [INFO]    {
[0000-00-00 00:00:00.000+0000] [INFO]        featureEnabled: false
[0000-00-00 00:00:00.000+0000] [INFO]        fn: [Function: fn]
[0000-00-00 00:00:00.000+0000] [INFO]        list: [ '1', '2', '3' ]
[0000-00-00 00:00:00.000+0000] [INFO]        nested: {
[0000-00-00 00:00:00.000+0000] [INFO]            level: {
[0000-00-00 00:00:00.000+0000] [INFO]                must: object
[0000-00-00 00:00:00.000+0000] [INFO]            }
[0000-00-00 00:00:00.000+0000] [INFO]        }
[0000-00-00 00:00:00.000+0000] [INFO]        noValue: {
[0000-00-00 00:00:00.000+0000] [INFO]            provided: null
[0000-00-00 00:00:00.000+0000] [INFO]            unprovided: undefined
[0000-00-00 00:00:00.000+0000] [INFO]        }
[0000-00-00 00:00:00.000+0000] [INFO]        numeric: 3.141592653589793
[0000-00-00 00:00:00.000+0000] [INFO]        string: 'complex'
[0000-00-00 00:00:00.000+0000] [INFO]        transformer: TextTransformer
[0000-00-00 00:00:00.000+0000] [INFO]    }
"
`;

exports[`TextTransformer format custom level text mask, will simply output get outputtet as is 1`] = `
"[NOW] [text] [INFO]    message
"
`;

exports[`TextTransformer format formatting errors without stack does not fail 1`] = `
"[NOW] [text] [ERROR]   Error: Error: no stack
"
`;

exports[`TextTransformer format unknown numeric level mask, without custom level text, will output UNKNOWN 1`] = `
"[NOW] [text] [INFO]    message
"
`;

exports[`TextTransformer formatTypes formatting errors without stack without breaking 1`] = `
"{
    errors: [ Without stack, With stack
        Error: With stack
at ... Mock stack trace ]
}"
`;

exports[`TextTransformer formatTypes named functions get outputted with name suffix 1`] = `
"{
    fns: [ Function ]
    withName: [Function: withName]
}"
`;
